### PR TITLE
Add type definitions for karma-webpack

### PR DESCRIPTION
**Pull request by mtraynham**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [X] The package does not provide its own types, and you can not add them.
- [X] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [X] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [X] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/karma-webpack/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// Type definitions for karma-webpack 2.0
+// Project: http://github.com/webpack/karma-webpack
+// Definitions by: Matt Traynham <https://github.com/mtraynham>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+import * as m from 'karma';
+import webpack = require('webpack');
+import webpackDevMiddleware = require('webpack-dev-middleware');
+
+declare module 'karma' {
```
Comment (plantain-00) [inline types/karma-webpack/index.d.ts]: It seems not right to declare module `karma` in `karma-webpack`

Comment (mtraynham) [reply types/karma-webpack/index.d.ts]: I kind of agree that it seems a bit weird, but this is basically an extension of the karma, or rather a module plugin. It appends two new properties to the Karma options.

It was started with the template `dts-gen --dt --name karma-webpack --template module-plugin`, which generates a [similar template](https://github.com/Microsoft/dts-gen/blob/master/templates/module-plugin.d.ts#L2).  It first imports the entire library you are adding to, and then declares a merged module to add new properties to.

**Comment by ryan-gustafson** [comment]
Is there a working example of this usage somewhere, outside of the test?

I'm struggling with getting the declaration merge to take affect in my `karma.conf.ts`.  The definition from `@types/karma` is located, but not merged with that from `@types/karma-webpack`, for whatever reason.  Thus the compiler complains `'webpack' does not exist in type 'ConfigOptions'`.

I can kludge around the problem with a type assertion, but that pretty much negates the benefits of typing...

**Comment by mtraynham** [comment]
Yeah the template for these add-on libraries is a bit strange.  I believe Typescript is supposed to resolve all relevant .d.ts files even if they aren't imported.  My local config looks like this:

```ts
import {karma} from './webpack.config';
import {Config} from 'karma';

export default function (config: Config): void {
    config.set({
        files: [
            'src/index.spec.ts'
        ],
        browsers: [
            'ChromeHeadless'
        ],
        singleRun: true,
        frameworks: [
            'jasmine'
        ],
        mime: {
            'text/x-typescript': ['ts', 'tsx']
        },
        preprocessors: {
            'src/index.spec.ts': ['webpack', 'sourcemap']
        },
        webpack: karma,
        webpackMiddlewareOptions: {
            noInfo: true
        },
        reporters: [
            'spec'
        ]
    });
}
```

If this doesn't work, what happens if you append `import 'karma-webpack';` to the top of the karma config?

**Comment by ryan-gustafson** [comment]
I'm on Typescript 2.4.2, a quick bump to 2.6.1 didn't seem to do anything, for what that's worth.  Strangely, `import 'karma-webpack';` seems to do the trick.  Could there be something else in your usage causing that typing information to be made available?  Or conversely, what's going in my usage?

Now I need to address the `karma-coverage` and `karma-remap-coverage` sections of the config options, but those are different, but equivalent, problems.

Curiously, `@types/karma-coverage` uses namespaces, but instead of exporting an equivalent `ConfigOptions` for merging, it extends the core `@types/karma` interface and exports that _new_ interface, so I must `import * from 'karma-coverage';` to locate it that version.  Using extension is a _greedy_ strategy though, it generally only works well once.

I cannot locate typing for `karma-remap-coverage`, but I was able to inline a module declaration to patch it up, for now.

There needs to be a strategy adopted for handling declaration merging which allows all the `karma` related configuration options to play nice together.  The module declaration approach seems to works well, allowing multiple independent extensions to be applied.

**Comment by mtraynham** [comment]
@ryan-gustafson , for reference, I created these typings by following the [dts-gen](https://github.com/Microsoft/dts-gen) [module-plugin template](https://github.com/Microsoft/dts-gen/blob/master/templates/module-plugin.d.ts), if you intend to change `karma-coverage` or add `karma-remap-coverage`.

It's a pretty simple process to create them, and here would be the workflow to do so:
```sh
npm install -g dts-gen
git clone https://github.com/<your fork>/DefinitelyTyped.git
cd DefinitelyTyped
dts-gen --dt --name karma-coverage --template module-plugin --overwrite
dts-gen --dt --name karma-remap-coverage --template module-plugin
```
Then you would go about changing the definitions as intended.

**Pull request merged**